File: common/driver.c
/* sec-ladder shared C driver helpers -- see common/driver.h. */
#include "driver.h"

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char *slb_arg_path(int argc, char **argv)
{
    if (argc != 2) {
        fprintf(stderr, "usage: %s <input-file>\n", argc > 0 ? argv[0] : "prog");
        exit(SLB_EXIT_USAGE);
    }
    return argv[1];
}

static uint64_t slb_le64(const unsigned char *p)
{
    return (uint64_t)p[0] | ((uint64_t)p[1] << 8) | ((uint64_t)p[2] << 16) |
           ((uint64_t)p[3] << 24) | ((uint64_t)p[4] << 32) | ((uint64_t)p[5] << 40) |
           ((uint64_t)p[6] << 48) | ((uint64_t)p[7] << 56);
}

void slb_load(const char *path, slb_input *out)
{
    unsigned char header[16];
    FILE *f = fopen(path, "rb");
    if (!f) {
        fprintf(stderr, "slb: cannot open %s\n", path);
        exit(SLB_EXIT_OPEN);
    }
    if (fread(header, 1, sizeof header, f) != sizeof header) {
        fprintf(stderr, "slb: %s: short header\n", path);
        fclose(f);
        exit(SLB_EXIT_HEADER);
    }
    out->n_iters = slb_le64(header);
    out->payload_len = slb_le64(header + 8);

    /* Read exactly payload_len bytes and insist they were all there. This is the
     * "length field larger than the payload" adversarial case; a sloppy driver
     * would trust the field and read uninitialised memory. */
    out->payload = NULL;
    if (out->payload_len > 0) {
        long here, end;
        size_t want = (size_t)out->payload_len;
        if ((uint64_t)want != out->payload_len) { /* 32-bit size_t guard */
            fprintf(stderr, "slb: %s: payload_len %" PRIu64 " exceeds size_t\n", path,
                    out->payload_len);
            fclose(f);
            exit(SLB_EXIT_TRUNCATED);
        }
        /* Check the declared length against the bytes actually present BEFORE
         * allocating. Rust's driver::load does not allocate on the declared
         * length either (it read_to_end's and compares), so without this the two
         * languages diverge on a huge payload_len: C would fail the malloc and
         * exit SLB_EXIT_NOMEM (6) where Rust exits SLB_EXIT_TRUNCATED (5), and
         * an adversarial input would look like a rung difference when it is a
         * driver difference. (TASK_002_REVIEW, minors.) */
        here = ftell(f);
        if (here >= 0 && fseek(f, 0, SEEK_END) == 0) {
            end = ftell(f);
            if (fseek(f, here, SEEK_SET) != 0) {
                fprintf(stderr, "slb: %s: cannot seek\n", path);
                fclose(f);
                exit(SLB_EXIT_TRUNCATED);
            }
            if (end >= 0 && (uint64_t)(end - here) < out->payload_len) {
                fprintf(stderr, "slb: %s: payload_len %" PRIu64 " exceeds file size\n",
                        path, out->payload_len);
                fclose(f);
                exit(SLB_EXIT_TRUNCATED);
            }
        }
        out->payload = (unsigned char *)malloc(want);
        if (!out->payload) {
            fprintf(stderr, "slb: %s: out of memory for %zu payload bytes\n", path, want);
            fclose(f);
            exit(SLB_EXIT_NOMEM);
        }
        if (fread(out->payload, 1, want, f) != want) {
            fprintf(stderr, "slb: %s: payload_len %" PRIu64 " exceeds file size\n", path,
                    out->payload_len);
            fclose(f);
            exit(SLB_EXIT_TRUNCATED);
        }
    }
    fclose(f);
}

uint64_t *slb_head_u64_body(const slb_input *in, uint64_t *head, size_t *n_body)
{
    size_t nwords = (size_t)(in->payload_len / 8);
    size_t i;
    uint64_t *body;

    if (nwords == 0) {
        *head = 0;
        *n_body = 0;
        return NULL;
    }
    *head = slb_le64(in->payload);
    *n_body = nwords - 1;
    if (*n_body == 0)
        return NULL;
    body = (uint64_t *)malloc(*n_body * sizeof *body);
    if (!body) {
        fprintf(stderr, "slb: out of memory for %zu body words\n", *n_body);
        exit(SLB_EXIT_NOMEM);
    }
    for (i = 0; i < *n_body; i++)
        body[i] = slb_le64(in->payload + 8 + 8 * i);
    return body;
}

unsigned char *slb_head2_u64_bytes(const slb_input *in, uint64_t *h0, uint64_t *h1,
                                   size_t *n_body)
{
    size_t len = (size_t)in->payload_len;
    unsigned char *body;

    if (in->payload_len < 16) {
        *h0 = 0;
        *h1 = 0;
        *n_body = 0;
        return NULL;
    }
    *h0 = slb_le64(in->payload);
    *h1 = slb_le64(in->payload + 8);
    *n_body = len - 16;
    if (*n_body == 0)
        return NULL;
    body = (unsigned char *)malloc(*n_body);
    if (!body) {
        fprintf(stderr, "slb: out of memory for %zu body bytes\n", *n_body);
        exit(SLB_EXIT_NOMEM);
    }
    memcpy(body, in->payload + 16, *n_body);
    return body;
}

unsigned char *slb_zeroed(uint64_t cap)
{
    unsigned char *p;
    if (cap == 0 || cap > SLB_MAX_CAP) {
        fprintf(stderr, "slb: destination capacity %" PRIu64 " out of range (1..%" PRIu64 ")\n",
                cap, (uint64_t)SLB_MAX_CAP);
        exit(SLB_EXIT_CAP);
    }
    p = (unsigned char *)calloc((size_t)cap, 1);
    if (!p) {
        fprintf(stderr, "slb: out of memory for %" PRIu64 " destination bytes\n", cap);
        exit(SLB_EXIT_NOMEM);
    }
    return p;
}

void slb_emit(uint64_t acc) { printf("%" PRIu64 "\n", acc); }

File: common/driver.rs
//! sec-ladder shared Rust driver helpers -- the line-for-line mirror of
//! common/driver.c.
//!
//! What lives here: reading the input file (`.memory/02-bench-rules.md` format),
//! decoding the payload, and printing the checksum. What deliberately does NOT
//! live here: the driver *loop*. Rung 5's loop must sit inside `verus!` so the
//! kernel call site is verified (`.memory/02-bench-rules.md`, rule 2), and a
//! shared loop cannot be both plain Rust and Verus. So every rung carries its
//! own copy of the loop between `SLB-DRIVER-BEGIN` / `SLB-DRIVER-END` markers
//! and `harness/check.py` diffs the copies.
//!
//! Pulled in with `#[path = "../../common/driver.rs"] mod driver;`. In
//! `verus.rs` that `mod` sits outside `verus! { .. }`, so Verus treats it as
//! external (external-by-default); `verus.rs` reaches it through one small
//! `#[verifier::external_body]` wrapper whose TCB cost is tallied in NOTES.md.
//!
//! Exit codes and stderr text are part of the contract: the `adversarial`
//! inputs compare rung behaviour, so a divergence here would look like a
//! finding when it is really a driver bug.

#![allow(dead_code)]

use std::io::Read;

pub const EXIT_USAGE: i32 = 2;
pub const EXIT_OPEN: i32 = 3;
pub const EXIT_HEADER: i32 = 4;
pub const EXIT_TRUNCATED: i32 = 5;

pub struct Input {
    pub n_iters: u64,
    pub payload_len: u64,
    pub payload: Vec<u8>,
}

fn die(code: i32, msg: &str) -> ! {
    eprintln!("{}", msg);
    std::process::exit(code)
}

fn le64(b: &[u8]) -> u64 {
    let mut w = [0u8; 8];
    w.copy_from_slice(&b[..8]);
    u64::from_le_bytes(w)
}

/// argv[1], or exit(EXIT_USAGE).
pub fn arg_path() -> String {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        let prog = args.first().cloned().unwrap_or_else(|| "prog".to_string());
        die(EXIT_USAGE, &format!("usage: {} <input-file>", prog));
    }
    args[1].clone()
}

/// Read `path`, or exit with one of the codes above.
pub fn load(path: &str) -> Input {
    let mut f = match std::fs::File::open(path) {
        Ok(f) => f,
        Err(_) => die(EXIT_OPEN, &format!("slb: cannot open {}", path)),
    };
    let mut header = [0u8; 16];
    if f.read_exact(&mut header).is_err() {
        die(EXIT_HEADER, &format!("slb: {}: short header", path));
    }
    let n_iters = le64(&header[0..8]);
    let payload_len = le64(&header[8..16]);

    // Read exactly payload_len bytes and insist they were all there: this is the
    // "length field larger than the payload" adversarial case.
    let mut payload = Vec::new();
    if payload_len > 0 {
        let want = payload_len as usize;
        if want as u64 != payload_len {
            die(
                EXIT_TRUNCATED,
                &format!("slb: {}: payload_len {} exceeds usize", path, payload_len),
            );
        }
        // read_to_end rather than a `want`-sized zeroed buffer: allocating on an
        // attacker-controlled length would let adversarial-shortlen OOM us
        // before we ever get to reject it.
        if f.read_to_end(&mut payload).is_err() || payload.len() < want {
            die(
                EXIT_TRUNCATED,
                &format!("slb: {}: payload_len {} exceeds file size", path, payload_len),
            );
        }
        payload.truncate(want);
    }
    Input { n_iters, payload_len, payload }
}

/// Decode a byte payload as little-endian u64s, ignoring a trailing partial word.
/// Mirrors `slb_u64s` in common/driver.c and `slb.u64s` in common/slb.py.
pub fn le_u64s(payload: &[u8]) -> Vec<u64> {
    let n = payload.len() / 8;
    let mut v = Vec::with_capacity(n);
    for i in 0..n {
        v.push(le64(&payload[8 * i..8 * i + 8]));
    }
    v
}

/// Split the payload into (head word, remaining words). Mirrors
/// `slb_head_u64_body` in common/driver.c. An empty payload yields (0, vec![]).
pub fn head_u64_body(input: &Input) -> (u64, Vec<u64>) {
    let mut w = le_u64s(&input.payload);
    if w.is_empty() {
        return (0, Vec::new());
    }
    let head = w[0];
    w.remove(0);
    (head, w)
}

/// Print the checksum as a decimal u64 and a newline. Nothing else on stdout.
pub fn emit(acc: u64) {
    println!("{}", acc);
}
